client/src/main/java/io/pravega/client/tables/KeyValueTable.java

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.tables;
+
+import io.pravega.common.util.AsyncIterator;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines all operations that are supported on a Key-Value Table.
+ * <p>
+ * A Key-Value Table is a distributed Key-Value Store that indexes Entries by Keys. It uses Table Segments (non-distributed
+ * Key-Value Store backed by a single Pravega Segment) as the fundamental storage primitive and provides a unified view
+ * of all Table Segments involved. Each {@link TableKey} is hashed to a Table Partition which may be represented by one
+ * or more Table Segments (depending on the Key-Value Table configuration chosen when it was created). Such partitioning
+ * enables the Key-Value Table to be distributed across the Pravega cluster but also introduces some constraints for
+ * certain operations (such as multi-key/entry atomic updates). See below for details.
+ * <p>
+ * Key Families are used to group related Keys together in the same Table Partition, which allows multiple
+ * keys/entries belonging to the same Key Family to be updated/removed atomically.
+ * <ul>
+ * <li> Multiple Keys/Entries in the same Key Family can be updated or removed atomically (either all at once or none).
+ * <li> Iterating through all Keys/Entries in the same Key Family is possible.
+ * <li> The same Key may exist in multiple Key Families or even not be associated with any Key Family at all. Such keys
```

Comment:
This concept is already complex enough without the additional possibility of some keys not being in a family. I think we can just make it mandatory. If users want them to be random they can just pass the same string twice.

Reply:
What I am concerned is that if users absolutely do not care about Key families, they will just do something like `put("", key, value)`, which will essentially group all keys into the same TableSegment, so there's no distribution done.

I am in favor of keeping both around.

Reply:
Perhaps we can then hide this inside of the key object itself. Have a Key type that we take which has an optional family field. That keeps this API clean, and there is no risk of misuse.

Reply:
That will cause trouble with batch updates and removals. Having the family separate from the keys forces the user to think that the update affects a single key family. Being able to mix keys from different key families together will not be supported, and we'll need to throw a runtime exception, which I think is worse than it is now.